### PR TITLE
[lua] [sql] Implement Elel's spawn conditions

### PR DESCRIPTION
**Pull request by WinterSolstice8**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Implement's Elel's specific spawn conditions (night, dark/double dark weather)

One commit is salvaged from ASB, though I re-made it because it was one giant commit making a ton of NM updates that aren't compatible and/or I don't want to test right now.

## Steps to test these changes

See Elel only spawn at night during weather or when weather changes.

### CODE REVIEW

**Comment by WinterSolstice8** [comment]
This is a little hard to test -- !setweather GLOOM and DARKNESS at night does the job. 
Any input on this would be nice, it seems to work but it feels hacky.
This kind of logic probably needs to hit King Vineragoon among others

**Comment by TeoTwawki** [comment]
> This is a little hard to test -- !setweather GLOOM and DARKNESS at night does the job. Any input on this would be nice, it seems to work but it feels hacky. This kind of logic probably needs to hit King Vineragoon among others

All the weather pops would work better the way elementals work, if it wasn't for the function that enables/disables spawning ignoring their other restrictions. Same issue for night pops - disallow gets set false, and they are allowed to pop - even when their spawn type says not to. I had to do something hacky to some night only spawns recently so don't worry about it being hacky here.

Is vinegaroon's issue that he pops too soon after weather change? Because I've not seen anything reported or had an issue spawning/despawning him so that's all I can think of being off.

File: scripts/zones/Sacrarium/IDs.lua [hunk 1]
```diff
@@ -46,6 +46,7 @@ zones[xi.zone.SACRARIUM] =
             { 16892074, { [xi.race.GALKA]    = true } },                            -- Indich (galka)
         },
         OLD_PROFESSOR_MARISELLE = 16891970,
+        ELEL                    = 16891948,
```
Comment (TeoTwawki) [inline scripts/zones/Sacrarium/IDs.lua]: we have that handy get-by-name stuff now so that we don't need the static ID.

**Comment by WinterSolstice8** [comment]
re KV;

If weather is active when it can spawn, it won't because it has to change _to_ single or double weather, checking it on zone tick in addition to this will cure that edge case. AFAIK KV can spawn any time weather is up, not just when it changes
```
zoneObject.onZoneWeatherChange = function(weather)
    local kvMob = GetMobByID(ID.mob.KING_VINEGARROON)

    if
        kvMob:getCurrentAction() == xi.act.DESPAWN and
        (weather == xi.weather.DUST_STORM or weather == xi.weather.SAND_STORM)
    then
        kvMob:spawn()
    elseif
        kvMob:getCurrentAction() == xi.act.ROAMING and
        weather ~= xi.weather.DUST_STORM and
        weather ~= xi.weather.SAND_STORM
    then
        DespawnMob(ID.mob.KING_VINEGARROON)
    end
end
```

**Comment by TeoTwawki** [comment]
> If weather is active when it can spawn, it won't because it has to change to single or double weather, checking it on zone tick in addition to this will cure that edge case. AFAIK KV can spawn any time weather is up, not just when it changes

I actually thought we had something happening on game hr or game day (not ideal) to pseudo cover that, but I see we do not currently.

File: scripts/zones/Sacrarium/IDs.lua (hunk 1 above)
Comment (WinterSolstice8) [reply scripts/zones/Sacrarium/IDs.lua]: done

File: scripts/zones/Sacrarium/Zone.lua [hunk 2]
```diff
@@ -62,4 +62,47 @@ end
 zoneObject.onEventFinish = function(player, csid, option, npc)
 end
 
+zoneObject.onZoneWeatherChange = function(weather)
+    local elel = GetMobByID(ID.mob.ELEL)
+    local vanadielHour = VanadielHour()
+
+    if
+        elel and
+        elel:getZone():getLocalVar('elelQueued') == 0 and -- Why doesn't onZoneWeatherChange contain the zone object...?
```
Comment (TeoTwawki) [inline scripts/zones/Sacrarium/Zone.lua:71]: man if I had a nickel every time I ran into that and asked the same thing...

I'd have 2. which isn't a lot but annoying that it happened twice right?

**Review by TeoTwawki — APPROVED** [review]
lgtm, till we have better handling of interactions between spawn types and disallow functionality

**Review by EpicTaru — APPROVED** [review]

**Review by Xaver-DaRed — APPROVED** [review]

**Pull request merged**